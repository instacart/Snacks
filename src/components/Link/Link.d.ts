import * as React from 'react'
import { RadiumStyles, ElementAttributes } from '../..'
import { WithThemeInjectedProps, ApplyWithTheme } from '../../styles/themer/withTheme'

export interface LinkProps extends WithThemeInjectedProps {
  /** The link's text content. */
  children?: React.ReactNode

  /** `href` attribute for the anchor tag. */
  href?: string

  /** Callback fired when the link is clicked. */
  onClick(e: React.MouseEvent<HTMLAnchorElement>, props: LinkProps): void

  /** Any additonal props */
  elementAttributes?: ElementAttributes<React.ComponentPropsWithoutRef<'a'>>

  style?: RadiumStyles
}

declare const Link: ApplyWithTheme<React.ComponentType<LinkProps>>

export default Link
