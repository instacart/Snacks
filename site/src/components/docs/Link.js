import React, {Fragment} from 'react'
import Playground from '../Playground'

export default () => (

<Fragment>
The `Link` component is valuable mainly because it's hooked up to Themer.
The link's resting/hover colors will match the current theme's action and actionHover values.

  <Playground>
    {`
  <div>
    <div style={{marginBottom: '16px'}}>
      <Link href="/radish">Normal link</Link>
    </div>

    <div style={{marginBottom: '16px'}}>
      <Link href="/radish" onClick={(e, props) => { e.preventDefault() }}>With an onClick callback</Link>
    </div>

    <div style={{marginBottom: '16px'}}>
      <Link href="/radish" onClick={(e, props) => { e.preventDefault() }} elementAttributes={{'aria-label': 'Foo'}}>With extra attributes like an aria-label</Link>
    </div>

    <div style={{marginBottom: '16px'}}>
      <Link href="/radish" onClick={(e, props) => { e.preventDefault() }} style={{textDecoration: 'underline', fontWeight: 600}}>With added styles</Link>
    </div>
  </div>
  `}
  </Playground>
</Fragment>

)
