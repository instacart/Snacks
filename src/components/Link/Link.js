import React from 'react'
import PropTypes from 'prop-types'
import Radium from 'radium'
import withTheme from '../../styles/themer/withTheme'
import { themePropTypes } from '../../styles/themer/utils'

const noop = () => {} // eslint-disable-line no-empty-function

const getStyles = props => {
  const { action, actionHover } = props.snacksTheme.colors

  return {
    cursor: 'pointer',
    textDecoration: 'none',
    color: action,

    ':hover': {
      color: actionHover,
      textDecoration: 'underline'
    }
  }
}

const Link = props => {
  const styles = getStyles(props)

  return (
    <a
      href={props.href}
      style={[styles, props.style]}
      onClick={e => {
        props.onClick(e, props)
      }}
      {...props.elementAttributes}
    >
      {props.children}
    </a>
  )
}

Link.propTypes = {
  /** `href` attribute for the anchor tag. */
  href: PropTypes.string,

  /** Callback fired when the link is clicked. */
  onClick: PropTypes.func,

  /** The link's text content. */
  children: PropTypes.node,

  /** Optional styles. */
  style: PropTypes.object,

  /** Any addional props. */
  elementAttributes: PropTypes.object,

  /** Snacks theme attributes provided by `Themer` */
  snacksTheme: themePropTypes
}

Link.defaultProps = {
  href: '#',
  onClick: noop
}

export default withTheme(Radium(Link))
