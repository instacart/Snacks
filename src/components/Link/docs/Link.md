Links!

```js
<div>
  <div style={{marginBottom: '16px'}}>
    <Link href="/radish">Normal link</Link>
  </div>

  <div style={{marginBottom: '16px'}}>
    <Link href="/radish" onClick={(e, props) => { e.preventDefault() }}>With an onClick callback</Link>
  </div>

  <div style={{marginBottom: '16px'}}>
    <Link href="/radish" onClick={(e, props) => { e.preventDefault() }} elementAttributes={{'aria-label': 'Foo'}}>With extra attributes like an aria-label</Link>
  </div>

  <div style={{marginBottom: '16px'}}>
    <Link href="/radish" onClick={(e, props) => { e.preventDefault() }} style={{textDecoration: 'underline', fontWeight: 600}}>With added styles</Link>
  </div>
</div>
```
