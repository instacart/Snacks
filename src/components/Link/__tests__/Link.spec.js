import Link from '../Link'
import React from 'react'
import renderer from 'react-test-renderer'
import { StyleRoot } from 'radium'
import { mount } from 'enzyme'
import { spy, match } from 'sinon'

fdescribe('Link', () => {
  it('renders without throwing', () => {
    const tree = renderer
      .create(
        <StyleRoot>
          <Link>HI</Link>
        </StyleRoot>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly when style overrides are provided', () => {
    const tree = renderer
      .create(
        <StyleRoot>
          <Link style={{ textDecoration: 'underline' }}>HI</Link>
        </StyleRoot>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('applies the elementAttributes prop correctly', () => {
    const tree = renderer
      .create(
        <StyleRoot>
          <Link elementAttributes={{ 'aria-label': 'foo' }}>HI</Link>
        </StyleRoot>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('fires the onClick callback with the correct arguments', () => {
    const onClick = spy()
    const props = {
      onClick,
      href: 'broccoli'
    }
    const wrapper = mount(
      <StyleRoot>
        <Link {...props}>HI</Link>
      </StyleRoot>
    )

    wrapper.find('a').simulate('click')

    const expectedProps = onClick.getCall(0).args[1]
    expect(props.onClick).toEqual(expectedProps.onClick)
    expect(props.href).toEqual(expectedProps.href)
  })
})
