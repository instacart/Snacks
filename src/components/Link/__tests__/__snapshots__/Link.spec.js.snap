// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link applies the elementAttributes prop correctly 1`] = `
<div
  data-radium={true}
>
  <a
    aria-label="foo"
    data-radium={true}
    href="#"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "color": "#43B02A",
        "cursor": "pointer",
        "textDecoration": "none",
      }
    }
  >
    HI
  </a>
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": "",
      }
    }
  />
</div>
`;

exports[`Link renders correctly when style overrides are provided 1`] = `
<div
  data-radium={true}
>
  <a
    data-radium={true}
    href="#"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "color": "#43B02A",
        "cursor": "pointer",
        "textDecoration": "underline",
      }
    }
  >
    HI
  </a>
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": "",
      }
    }
  />
</div>
`;

exports[`Link renders without throwing 1`] = `
<div
  data-radium={true}
>
  <a
    data-radium={true}
    href="#"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "color": "#43B02A",
        "cursor": "pointer",
        "textDecoration": "none",
      }
    }
  >
    HI
  </a>
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": "",
      }
    }
  />
</div>
`;
